### PR TITLE
cpu/cortexm: "free" SVC

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

Currently, the Cortex-M "supervisor call exception" SVC (think system call entry point) is completely blocked by being defined to be tail-called by PendSV, in order to call `sched_run()` and restore the context from `sched_active_thread`. This makes it unusable for any other use but `cpu_switch_context_exit()`.

This PR refactors `PendSV` to include the previous SVC routine (run scheduler, restore thread context), but before that, only conditionally stores the current context only if `sched_active_thread` is non-null.

An SVC dispatch function is added. SVC 1 maps to the original function by handing over to PendSV.

This is the first step to implement system calls using SVC, which can now be used in combination with the MPU to provide some memory protected "user space" that still interacts with the kernel.

Performance wise, context switches are slightly faster on Cortex-M3 onwards, and slightly slower on CM0(+), but both times less than 0.5%. Crude measuring shows ~50b increase in code size.

I've tested this only on particle-xenon (Cortex-M4F) and nucleo-f072rb (Cortex-M0+).

### Testing procedure

Being quite central with a lot of opportunities for weird error cases, this should get somewhat extensive review and testing, especially anything involving context switching on different Cortex-M.

### CODE REVIEW

**Comment by kaspar030** [comment]
I've pushed a commit that makes the dispatching optional on selecting the module `cortexm_svc`, basically restoring the previous default behavior.

File: cpu/cortexm_common/thread_arch.c [hunk 1]
```diff
@@ -366,3 +367,75 @@ void __attribute__((naked)) __attribute__((used)) isr_svc(void) {
...
+     * number is the lower byte of the instruction.
+     */
+    unsigned int svc_number = ((char *)svc_args[6])[-2];
+
+    switch (svc_number) {
+        case 1: /* SVC number used by cpu_switch_context_exit */
+            SCB->ICSR |= SCB_ICSR_PENDSVSET_Msk;
+            break;
+        default:
+            //DEBUG("svc: unhandled SVC #%u\n", svc_number);
```
Comment (fjmolinas) [inline cpu/cortexm_common/thread_arch.c:430]: Either remove `DEBUG` or add 
```
#define ENABLE_DEBUG    (0)
#include "debug.h"
```

File: makefiles/pseudomodules.inc.mk [hunk 2]
```diff
@@ -9,6 +9,7 @@ PSEUDOMODULES += conn_can_isotp_multi
 PSEUDOMODULES += cord_ep_standalone
 PSEUDOMODULES += core_%
 PSEUDOMODULES += cortexm_fpu
+PSEUDOMODULES += cortexm_svc
```
Comment (fjmolinas) [inline makefiles/pseudomodules.inc.mk:12]: This should be provided in `cpu/cortexm_common/Makefile.features`

File: cpu/cortexm_common/thread_arch.c [hunk 3]
```diff
@@ -286,9 +286,17 @@ void __attribute__((naked)) __attribute__((used)) isr_pendsv(void) {
     /* save context by pushing unsaved registers to the stack */
     /* {r0-r3,r12,LR,PC,xPSR,s0-s15,FPSCR} are saved automatically on exception entry */
     ".thumb_func                      \n"
+
+    /* skip context saving if sched_active_thread == NULL */
+    "ldr    r1, =sched_active_thread  \n" /* r1 = &sched_active_thread  */
+    "ldr    r1, [r1]                  \n" /* r1 = sched_active_thread   */
+    "cmp    r1, #0                    \n" /* if r1 == NULL:             */
+    "beq context_restore              \n" /*     goto context_restore   */
```
Comment (fjmolinas) [inline cpu/cortexm_common/thread_arch.c:294]: Skipping context switching depending on `sched_active_thread` should go into its own commit.

File: cpu/cortexm_common/thread_arch.c [hunk 4]
```diff
@@ -311,23 +320,15 @@ void __attribute__((naked)) __attribute__((used)) isr_pendsv(void) {
...
-    /* SVC handler entry point */
-    /* PendSV will continue here as well (via jump) */
-    ".thumb_func                      \n"
-    /* perform scheduling */
-    "bl     sched_run                 \n"
-    /* restore context and return from exception */
-    ".thumb_func                      \n"
     "context_restore:                 \n"
+
+    "bl     sched_run                 \n" /* perform scheduling */
```
Comment (fjmolinas) [inline cpu/cortexm_common/thread_arch.c:329]: `sched_run` should not be part of `context_restore` IMO, I understand this is because of the branch at L294, but to keep the same ordering I would rather have:

```suggestion
    "run_shed:                         \n"
    "bl     sched_run                 \n" /* perform scheduling */
    "context_restore:                 \n"
```

File: cpu/cortexm_common/thread_arch.c [hunk 5]
```diff
@@ -366,3 +367,75 @@ void __attribute__((naked)) __attribute__((used)) isr_svc(void) {
...
+    );
+#endif
+}
+
+void svc_dispatch(unsigned int *svc_args)
+{
+    /* svc_args contains:
+     * r0, r1, r2, r3, r12, r14, the return address and xPSR
+     * First argument (r0) is svc_args[0]
+     */
```
[the remark this answers is not in the document]
Comment (fjmolinas) [reply cpu/cortexm_common/thread_arch.c:416]: Is this function meant to be private?

File: cpu/cortexm_common/thread_arch.c [hunk 6]
```diff
@@ -366,3 +367,75 @@ void __attribute__((naked)) __attribute__((used)) isr_svc(void) {
...
+    "tst    lr, #4          \n" /* switch bit4(lr) == 1):   */
+    "ite    eq              \n"
+    "mrseq  r0, msp         \n" /* case 1: r0 = msp         */
+    "mrsne  r0, psp         \n" /* case 0: r0 = psp         */
+    "b      svc_dispatch    \n" /* return svc_dispatch()    */
+    );
+#endif
+}
+
+void svc_dispatch(unsigned int *svc_args)
```
Comment (fjmolinas) [inline cpu/cortexm_common/thread_arch.c:411]: Just nitpicking here, buy why `dispatch` and not `svc_handler`?

File: cpu/cortexm_common/thread_arch.c (hunk 1 above)
Comment (kaspar030) [reply cpu/cortexm_common/thread_arch.c:430]: ok

File: makefiles/pseudomodules.inc.mk (hunk 2 above)
Comment (kaspar030) [reply makefiles/pseudomodules.inc.mk:12]: yup

File: cpu/cortexm_common/thread_arch.c (hunk 3 above)
Comment (kaspar030) [reply cpu/cortexm_common/thread_arch.c:294]: ok, will re-shuffle when squashing

File: cpu/cortexm_common/thread_arch.c (hunk 4 above)
Comment (kaspar030) [reply cpu/cortexm_common/thread_arch.c:329]: hm, those are goto labels. Noone jumps to "run_shed", so we on't need it. Maybe "context_restore" could be named "context_select_and_restore" or something.

File: cpu/cortexm_common/thread_arch.c (hunk 6 above)
Comment (kaspar030) [reply cpu/cortexm_common/thread_arch.c:411]: Well, there are three "classes" involved:

1. the actual exception handler (`isr_svc()`)
2. something that determines what function to call (does dispatching depending on svc number, `svc_dispatch()`)
3. handlers for some svc numbers (we don't have those yet).

That's why it is called "dispatch". ;) makes sense?

Comment (fjmolinas) [reply cpu/cortexm_common/thread_arch.c:411]: Yep!

File: cpu/cortexm_common/thread_arch.c (hunk 4 above)
Comment (fjmolinas) [reply cpu/cortexm_common/thread_arch.c:329]: > hm, those are goto labels. Noone jumps to "run_shed", so we on't need it. Maybe "context_restore" could be named "context_select_and_restore" or something.

I would branch to it on L294 `"beq run_sched             \n" /*     goto context_restore   */`

I just don't like `sched_run` being part of `context_restore`. I would also be Ok with a name change.

Comment (kaspar030) [reply cpu/cortexm_common/thread_arch.c:332]: renamed it to "select_and_restore_context"

**Comment by fjmolinas** [comment]
Bench-marking `tests/bench_thread_yield_pingpong'`

|        core        | master |   pr   | pr + cortexm_svc  |
|:------------------:|:------:|:------:|:-----------------:|
|   m0+ (usb-kw41z)  |  96632 |  95862 |       95862       |
| m0 (nucle0-f091rc) |  85683 |  85076 |       85076       |
|   m3 (openmote-b)  |  70123 |  70741 |       70741       |
|   m3 (iotlab-m3)   | 161035 | 158550 |       158550      |
|   m4 (frdm-k64f)   | 150376 | 149254 |       149254      |
| m7 (nucleo-f767zi) | 291104 | 289156 |       289156      |

In general they are all running faster.

**Comment by fjmolinas** [comment]
`tests/thread_cooperation` still working correctly for the above mentioned boards, any other testing suggestion? I could run a subset of tests on many boards?

File: cpu/cortexm_common/thread_arch.c (hunk 4 above)
Comment (fjmolinas) [reply cpu/cortexm_common/thread_arch.c]: ok

File: cpu/cortexm_common/thread_arch.c [hunk 7]
```diff
@@ -366,3 +370,81 @@ void __attribute__((naked)) __attribute__((used)) isr_svc(void) {
     /* {r0-r3,r12,LR,PC,xPSR,s0-s15,FPSCR} are restored automatically on exception return */
     );
 }
+
+#ifdef MODULE_CORTEXM_SVC
+void svc_dispatch(unsigned int *svc_args);
```
Comment (fjmolinas) [inline cpu/cortexm_common/thread_arch.c:375]: How will this function be used? shouldn't this be in `cpu.h`?

Comment (kaspar030) [reply cpu/cortexm_common/thread_arch.c:375]: It is called by `isr_svc()`.

Comment (fjmolinas) [reply cpu/cortexm_common/thread_arch.c:375]: Sorry I didn't express my question correctly, I mean shouldn't this be private/static in that case? Maybe using `__attribute__((used))`?

**Comment by kaspar030** [comment]
> In general they are all running faster. Only openmote-b had a .8% increase.

Aehm, the benchmark counts context switch per second, so lower is better ;)

I'm seeing a slight improvement on nucleo-f401re.

File: cpu/cortexm_common/thread_arch.c (hunk 7 above)
Comment (kaspar030) [reply cpu/cortexm_common/thread_arch.c:375]: I've made it static!

**Comment by fjmolinas** [comment]
> Aehm, the benchmark counts context switch per second, so lower is better ;) edit sorry for adding my own confusion, higher numbers are better.

Well then iotlab-m3 is slower.

File: cpu/cortexm_common/thread_arch.c [hunk 8]
```diff
@@ -366,3 +370,81 @@ void __attribute__((naked)) __attribute__((used)) isr_svc(void) {
...
+    "b      svc_dispatch    \n" /* return svc_dispatch(r0)  */
+    );
+#else
+    __asm__ volatile (
+    ".thumb_func            \n"
+    "tst    lr, #4          \n" /* switch bit4(lr) == 1):   */
+    "ite    eq              \n"
+    "mrseq  r0, msp         \n" /* case 1: r0 = msp         */
+    "mrsne  r0, psp         \n" /* case 0: r0 = psp         */
+    "b      svc_dispatch    \n" /* return svc_dispatch()    */
```
Comment (fjmolinas) [inline cpu/cortexm_common/thread_arch.c:409]:
```suggestion
    "b     _svc_dispatch    \n" /* return svc_dispatch()    */
```

File: cpu/cortexm_common/thread_arch.c [hunk 9]
```diff
@@ -366,3 +370,81 @@ void __attribute__((naked)) __attribute__((used)) isr_svc(void) {
...
+    );
+#endif
+}
+
+static void __attribute__((used)) svc_dispatch(unsigned int *svc_args)
```
Comment (fjmolinas) [inline cpu/cortexm_common/thread_arch.c:414]:
```suggestion
static void __attribute__((used)) _svc_dispatch(unsigned int *svc_args)
```

File: cpu/cortexm_common/thread_arch.c [hunk 10]
```diff
@@ -366,3 +370,81 @@ void __attribute__((naked)) __attribute__((used)) isr_svc(void) {
     /* {r0-r3,r12,LR,PC,xPSR,s0-s15,FPSCR} are restored automatically on exception return */
     );
 }
+
+#ifdef MODULE_CORTEXM_SVC
+static void __attribute__((used)) svc_dispatch(unsigned int *svc_args);
```
Comment (fjmolinas) [inline cpu/cortexm_common/thread_arch.c:375]:
```suggestion
```

Comment (fjmolinas) [inline cpu/cortexm_common/thread_arch.c:375]: Not needed I believe
```suggestion
```

File: cpu/cortexm_common/thread_arch.c (hunk 8 above)
Comment (kaspar030) [reply cpu/cortexm_common/thread_arch.c:409]: done

File: cpu/cortexm_common/thread_arch.c (hunk 9 above)
Comment (kaspar030) [reply cpu/cortexm_common/thread_arch.c:414]: yup

File: cpu/cortexm_common/thread_arch.c (hunk 10 above)
Comment (kaspar030) [reply cpu/cortexm_common/thread_arch.c:375]: :+1:

**Comment by kaspar030** [comment]
Hm, I'm seeing different behavior gcc vs llvm, and gcc versions in our toolchain vs. 9.3.0, on both nucleo-f401re and samr21-xpro. llvm is always slightly slower on master (both current llvm and the one in our container). gcc in our container is slightly slower with this pr. current gcc is slightly faster with this pr. :-/

**Comment by fjmolinas** [comment]
> Hm, I'm seeing different behavior gcc vs llvm, and gcc versions in our toolchain vs. 9.3.0, on both nucleo-f401re and samr21-xpro. llvm is always slightly slower on master (both current llvm and the one in our container). gcc in our container is slightly slower with this pr. current gcc is slightly faster with this pr. :-/

When using `llvm` on `iotlab-m3` I get more context switches on master `176432` vs `173876` on this PR.

**Comment by kaspar030** [comment]
So, with llvm, PR is faster. With gcc, this PR is generally slower. It might get faster *with* PR on newer gcc releases.

As we don't have an actual use case ATM, I guess we have to wait. No need to slow down context switches for nothing.

File: cpu/cortexm_common/Makefile.features [hunk 11]
```diff
@@ -2,6 +2,7 @@ FEATURES_PROVIDED += arch_32bit
 FEATURES_PROVIDED += arch_arm
 FEATURES_PROVIDED += arch_cortexm
 FEATURES_PROVIDED += periph_pm
+FEATURES_PROVIDED += cortexm_svc
```
Comment (fjmolinas) [inline cpu/cortexm_common/Makefile.features:5]: Please also add in `makefiles/features_modules.inc.mk`

```
# select cpu_check_address pseudomodule if the corresponding feature is used
USEMODULE += $(filter cortexm_svc, $(FEATURES_USED))
```

**Comment by kaspar030** [comment]
@leandrolanzieri @jia200x can you take a look if I added the feature to Kconfig correctly?

**Comment by leandrolanzieri** [comment]
> @leandrolanzieri @jia200x can you take a look if I added the feature to Kconfig correctly?

Yes, LGTM!

File: cpu/cortexm_common/Makefile.features (hunk 11 above)
Comment (kaspar030) [reply cpu/cortexm_common/Makefile.features:5]: done

**Comment by kaspar030** [comment]
This is now a requirement for #14224. IMO that makes the slight performance drop for some compilers / MCUs acceptable...

**Comment by maribu** [comment]
> IMO that makes the slight performance drop for some compilers / MCUs acceptable...

Actually, now is the perfect time to add some features coming at the price of reduced performance: The `inline`-able IRQ API improved the performance so much, that the next release is still going to be faster than the current one. So users won't notice :smiling_imp:

**Comment by fjmolinas** [comment]
I would like to see a run of all tests here as well.

**Comment by fjmolinas** [comment]
Test failures are in esp32, so unrelated, I'll re-trigger without them

**Comment by fjmolinas** [comment]
No one has said anything for while so GO!

**Comment by kaspar030** [comment]
> No one has said anything for while so GO!

Thanks for reviewing!
